### PR TITLE
Some corner cases fixes

### PR DESCRIPTION
**Pull request by killme2008**
Main changes:

* Replicate logs before appliedIndex(in follower) to followers should be considered successfully.
* Provide an isolate thread pool to run replciator and rpc closures to prevent problems such as #426 
* Prevent creating too many block timers when RPC fails( status is not ok).
* Fixed replicate state field may be wrong in some situations.

### CODE REVIEW

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java [hunk 1]
```diff
@@ -910,6 +910,7 @@ static boolean continueSending(final ThreadId id, final int errCode) {
             // _wait_more_entries and no further logs would be replicated even if the
             // last_index of this followers is less than |next_index - 1|
             r.sendEmptyEntries(false);
+            r.blockTimer = null;
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:913]: 这行去掉的逻辑是什么？

Comment (killme2008) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:913]: 这是加，哪有去？

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java [hunk 2]
```diff
@@ -930,6 +931,11 @@ void block(final long startTimeMs, @SuppressWarnings("unused") final int errorCo
         // each individual error (e.g. we don't need check every
         // heartbeat_timeout_ms whether a dead follower has come back), but it's just
         // fine now.
+        if(this.blockTimer!=null) {
```
[the remark this answers is not in the document]
Comment (killme2008) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:934]: 嗯，先别看了，还需要补充测试，未完成

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java (hunk 1 above)
Comment (fengjiachun) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:913]: 擦，看错了

**Comment by killme2008** [comment]
@fengjiachun @ujjboy @zongtanghu 有空帮忙看下

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java [hunk 3]
```diff
@@ -1417,10 +1425,10 @@ private static boolean onAppendEntriesReturned(final ThreadId id, final Inflight
                 LOG.debug("Replicated logs in [{}, {}] to peer {}", r.nextIndex, r.nextIndex + entriesSize - 1,
                     r.options.getPeerId());
             }
-        } else {
-            // The request is probe request, change the state into Replicate.
-            r.state = State.Replicate;
         }
+
+        r.state = State.Replicate;
+        r.blockTimer = null;
```
Comment (zongtanghu) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:1431]: 这里是，只要Leader->Follower的RPC请求return回来ok，则都将replicator的blockTimer定时任务清理？其实也可以做下if的判断，如果不为null，再清理？

File: jraft-core/src/main/java/com/alipay/sofa/jraft/rpc/RpcUtils.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+     */
+    public static Future<?> runInThread(final Runnable runnable) {
+        return RPC_CLOSURE_EXECUTOR.submit(runnable);
+    }
+
+    /**
+     * Run closure with status in thread pool.
+     */
+    public static Future<?> runClosureInThread(final Closure done, final Status status) {
+    if (done == null) {
```
Comment (zongtanghu) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/rpc/RpcUtils.java:96]: 格式，缩进问题。

File: jraft-core/src/main/java/com/alipay/sofa/jraft/storage/impl/LogManagerImpl.java [hunk 5]
```diff
@@ -1008,6 +1008,8 @@ private boolean checkAndResolveConflict(final List<LogEntry> entries, final Stab
                 LOG.warn(
                     "Received entries of which the lastLog={} is not greater than appliedIndex={}, return immediately with nothing changed.",
                     lastLogEntry.getId().getIndex(), appliedIndex);
+                // Replicate old logs before appliedIndex should be considered successfully, response OK.
+                Utils.runClosureInThread(done);
```
Comment (zongtanghu) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/storage/impl/LogManagerImpl.java:1012]: 这块是没有替换成RpcUtils.runClosureInThread(done)?

**Comment by zongtanghu** [comment]
I have already put forward some code review suggestion. @killme2008

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java [hunk 6]
```diff
@@ -262,13 +263,13 @@ private static void notifyReplicatorStatusListener(final Replicator replicator,
...
-                            Utils.runInThread(() -> listener.onCreated(peer));
+                            RpcUtils.runInThread(() -> listener.onCreated(peer));
                             break;
                         case ERROR:
-                            Utils.runInThread(() -> listener.onError(peer, status));
+                          RpcUtils.runInThread(() -> listener.onError(peer, status));
                             break;
                         case DESTROYED:
-                            Utils.runInThread(() -> listener.onDestroyed(peer));
+                          RpcUtils.runInThread(() -> listener.onDestroyed(peer));
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:272]: 格式有问题

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java [hunk 7]
```diff
@@ -262,13 +263,13 @@ private static void notifyReplicatorStatusListener(final Replicator replicator,
                 try {
                     switch (event) {
                         case CREATED:
-                            Utils.runInThread(() -> listener.onCreated(peer));
+                            RpcUtils.runInThread(() -> listener.onCreated(peer));
                             break;
                         case ERROR:
-                            Utils.runInThread(() -> listener.onError(peer, status));
+                          RpcUtils.runInThread(() -> listener.onError(peer, status));
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:269]: 格式有问题

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java [hunk 8]
```diff
@@ -839,7 +840,7 @@ public static void waitForCaughtUp(final ThreadId id, final long maxMargin, fina
         final Replicator r = (Replicator) id.lock();
 
         if (r == null) {
-            Utils.runClosureInThread(done, new Status(RaftError.EINVAL, "No such replicator"));
+          RpcUtils.runClosureInThread(done, new Status(RaftError.EINVAL, "No such replicator"));
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:843]: 格式

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java [hunk 9]
```diff
@@ -921,7 +923,7 @@ static boolean continueSending(final ThreadId id, final int errCode) {
     }
 
     static void onBlockTimeout(final ThreadId arg) {
-        Utils.runInThread(() -> onBlockTimeoutInNewThread(arg));
+      RpcUtils.runInThread(() -> onBlockTimeoutInNewThread(arg));
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:926]: 格式

File: jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java [hunk 10]
```diff
@@ -930,6 +932,11 @@ void block(final long startTimeMs, @SuppressWarnings("unused") final int errorCo
         // each individual error (e.g. we don't need check every
         // heartbeat_timeout_ms whether a dead follower has come back), but it's just
         // fine now.
+        if(this.blockTimer != null) {
+          // already in blocking state,return immediately.
+          this.id.unlock();
+          return;
+        }
```
Comment (fengjiachun) [inline jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:939]: 缩进改为 2 个空格了？

Comment (killme2008) [reply jraft-core/src/main/java/com/alipay/sofa/jraft/core/Replicator.java:939]: 格式先不用管，看看代码

**Pull request merged**